rasterio/_base.pyx

```diff
@@ -444,7 +442,7 @@ cdef class DatasetBase:
         self._env.__exit__()
 
     def __dealloc__(self):
-        if self._hds != NULL:
+        if not self.closed:
```

Comment:
@groutr `__dealloc__`, like `__cinit__` shouldn't know anything about the class other than its extension members (like `_hds`).